x-pack/apm-server/sampling/processor.go

```diff
@@ -112,45 +164,61 @@ func (p *Processor) ProcessTransformables(ctx context.Context, events []transfor
 			events = events[:n-1]
 			i--
 		}
+		if stored {
+			atomic.AddInt64(&p.eventMetrics.stored, 1)
+		} else if drop {
+			// We only increment the "dropped" counter if
+			// we didn't also store the event, so we can
+			// track how many events are definitely dropped
+			// and indexing isn't just deferred until later.
+			//
+			// The counter does not include events that are
+			// implicitly dropped, i.e. stored and never
+			// indexed.
+			atomic.AddInt64(&p.eventMetrics.dropped, 1)
+		}
 	}
 	return events, nil
 }
 
-func (p *Processor) processTransaction(tx *model.Transaction) (bool, error) {
+func (p *Processor) processTransaction(tx *model.Transaction) (drop, stored bool, _ error) {
 	if tx.Sampled != nil && !*tx.Sampled {
 		// (Head-based) unsampled transactions are passed through
 		// by the tail sampler.
-		return false, nil
+		return false, false, nil
 	}
 
 	traceSampled, err := p.storage.IsTraceSampled(tx.TraceID)
 	switch err {
 	case nil:
 		// Tail-sampling decision has been made, index or drop the transaction.
 		drop := !traceSampled
-		return drop, nil
+		return drop, false, nil
 	case eventstorage.ErrNotFound:
 		// Tail-sampling decision has not yet been made.
 		break
 	default:
-		return false, err
+		return false, false, err
 	}
 
 	if tx.ParentID != "" {
 		// Non-root transaction: write to local storage while we wait
 		// for a sampling decision.
-		return true, p.storage.WriteTransaction(tx)
+		return true, true, p.storage.WriteTransaction(tx)
```

Comment:
Why is `drop: true` in this case? I see that in https://github.com/elastic/apm-server/pull/4346/files#diff-23ea3f45d5b6898582f132034e90a9e3da30d7345b4cbe4c5799237094e418cfR169 you are only using the `drop` information if `stored` is `false` but I would find it more intuitive to only set `drop: true` once it is really dropped. Maybe I am missing something, but I think changing it here would not have any side effects.

Reply:
"dropped" has a double meaning, which is a bit confusing. Here we're saying that we should drop the transaction from the list of events that are to be returned from ProcessTransformables, to prevent them from being indexed immediately.

Perhaps it would be more understandable if I inverted this return variable? i.e. instead of telling ProcessTransformables to drop the event, tell it if we want it to report the event immediately. Then we'll increment the dropped counter if and only if we're not reporting immediately, and we also did not store the event for later.

Reply:
If you don't mind changing, I believe it would make things more clear.

Reply:
I've pushed a change, please let me know if it's clear enough now.

Reply:
Thank you - the changes make it easier to understand